### DISCUSSION TITLE
A couple issues or some advice on coding with the circle SDK

### DISCUSSION THREAD

**Discussion by ArnoldUK** [rsta2/circle#360] [General]
Hi all,

Dabbling with the SDK and quite impressed with what can be achieved using the circle SDK.
I'm part way through porting some emulator code to interface with the circle SDK.  I've wrritten this small test program to update the framebuffer pixels but not quite sure my method is the correct and fastest way to do this. Below is my not so impressive code which is for testing purposes:
```
TShutdownMode CKernel::Run (void) {
	int scrWidth = m_Screen.GetWidth();
	int scrHeight = m_Screen.GetHeight();
        
       CBcmRandomNumberGenerator m_Random;
	CBcmFrameBuffer	*m_pFrameBuffer;
	//m_pFrameBuffer = new CBcmFrameBuffer (scrWidth, scrHeight, 8);
	m_pFrameBuffer = m_Screen.GetFrameBuffer();
	
	unsigned char cRed, cGreen, cBlue;
	
	TScreenColor *m_pBuffer;
	m_pBuffer = (TScreenColor *) (uintptr) m_pFrameBuffer->GetBuffer();

	int bufWidth = m_pFrameBuffer->GetWidth();
	int bufHeight = m_pFrameBuffer->GetHeight();
	int bufPitch = m_pFrameBuffer->GetPitch();
	
	int nPosY, nPosX, rCol;
	
	while(true) {
		m_Scheduler.Yield();
		m_Timer.MsDelay(1);
		
		cRed = m_Random.GetNumber () % 31;
		cGreen = m_Random.GetNumber () % 31;
		cBlue = m_Random.GetNumber () % 31;
		
		for (nPosY = 0; nPosY < bufHeight; nPosY++) {
			for (nPosX = 0; nPosX < bufWidth; nPosX++) {
				m_pBuffer[bufPitch * nPosY + nPosX] =  COLOR16(cRed, cGreen, cBlue);
			}
		}
	}
	
	return ShutdownHalt;
}
```

Also, what is the largest static array that can be created ?
I've had some issues with the kernel freezing when declaring global static arrays for example:
`uint32_t vgaframebuffer[1024][1024];`
`uint32_t memorybuffer[8192];`

I know creating large dynamic arrays is always preferred using malloc but I have no idea what is the largest static array that can be created using circle kernel for say the Raspberry 1 or 2, whithout exhausting the global heap?

Thanks for reading and any advice much appreciated. Sorry about the code above not formatting correctly but I did quote in code symbols.

**Comment by rsta2** [comment]
Hi, there is also a `m_Screen.SetPixel()` method, which would help to get rid of accessing the frame buffer directly, but if you are looking for the fastest way, your code should be OK. You do not need to call `m_Scheduler.Yield()`, if you to not have secondary tasks.

Circle kernel images can have a maximum size of 2 MBytes (including code and static data) by default. Static arrays must fit into this. If you want more, you can edit the `KERNEL_MAX_SIZE` system option in *include/circle/sysconfig.h* or add `DEFINE += -DKERNEL_MAX_SIZE=0x400000` to the *Config.mk* file for example for a maximum size of 4 MBytes.

Code snippets should be enclosed into \`\`\` on empty lines here.

**Comment by ArnoldUK** [comment]
Great, thanks rsta2 for approving my code to update the frame buffer. I'm aware of the SetPixel and GetPixel functions but at the moment this code is simply a test run before I add the actual finsished screen update code. The KERNEL_MAX_SIZE define I have read about in the docs but I thought that was for the actual kernel.img file size. I will add the define flag as you suggested and test my code again.

Oh, one question about the CBcmFrameBuffer , if you look at my code, I have commented out a line that creates a new CBcmFrameBuffer :
```
CBcmFrameBuffer	*m_pFrameBuffer;
//m_pFrameBuffer = new CBcmFrameBuffer (scrWidth, scrHeight, 8);
m_pFrameBuffer = m_Screen.GetFrameBuffer();
```
Is it better to create a new `CBcmFrameBuffer` or use the existing `CBcmFrameBuffer.GetFrameBuffer()` ?
The code runs fine either way I use the CBcmFrameBuffer but wondered what benefits if any.

Fixed my post snippets now, thanks for that too. Your replies have certainly helped!

**Reply by rsta2** [reply]
You are welcome. The class `CScreenDevice` already contains a member object of class `CBcmFrameBuffer`, so if you create another object of this class, you have two objects, which try to access the same frame buffer. This is not recommended and only works, because you haven't called the `Initialize()` method of the second instance.

**Comment by ArnoldUK** [comment]
Ok, that explains things enough for me for now. Thanks again.

**Comment by ArnoldUK** [comment]
While I still have this discussion open there is another issue that I'm not sure about and that is to do with threads.
In paricular using the `<pthread.h>` header file for `pthread_create` and `pthread_exit` methods. Would these still work as normal in my code if compiling with Circle SDK and arm-none-eabi compiler ?

I had noticed that the Circle SDK included the CScheduler and CTask classes for multitasking but I have not looked any further into these classes and if they would offer any improvement or stability to my code.

**Comment by rsta2** [comment]
The pthread support in *addon/linux* is very limited and has only be implemented in Circle to be able to port the accelerated graphics support. You have to include `<linux/pthread.h>` (not `<pthread.h>`) to try it, but I wouldn't recommend this.

The native cooperative task support in Circle is described [here](https://circle-rpi.readthedocs.io/en/45.0/subsystems/multitasking.html). Have a look at the [sample/19-tasks](https://github.com/rsta2/circle/tree/master/sample/19-tasks).

**Comment by ArnoldUK** [comment]
Brilliant! I really appreciate your help. I've got 80% of this working now and I will post a link to the finished project of  XT Emulator for RPi built with the Circle SDK.
